util/upload-cookbook.py

```diff
@@ -0,0 +1,241 @@
+                full_name, bucket_name, args.override, args.dryrun
+            )
+        )
+    else:
+        key = folder + "/" + full_name
+        try:
+            s3.upload_file(full_name, bucket_name, key, ExtraArgs={"ACL": "public-read"})
+
+            print("Successfully uploaded {0} to s3://{1}/{2}".format(full_name, bucket_name, key))
+        except ClientError as e:
+            print("Couldn't upload {0} to bucket s3://{1}/{2}".format(full_name, bucket_name, key))
+            _cp_error_array.add(region)
+            if e.response["Error"]["Code"] == "NoSuchBucket":
+                print("Bucket is not present.")
+
+            raise e
+
+
+def _create_s3_client(region):
+    reg_credentials = [c for c in credentials if c[0] == region]
+
+    if reg_credentials:
+        credential = reg_credentials[0]
+        credential_region = credential[0]
+        credential_endpoint = credential[1]
+        credential_arn = credential[2]
+        credential_external_id = credential[3]
+
+        try:
+            sts = boto3.client("sts", region_name=main_region, endpoint_url=credential_endpoint)
+
+            assumed_role_object = sts.assume_role(
+                RoleArn=credential_arn,
+                ExternalId=credential_external_id,
+                RoleSessionName=credential_region + "upload_cfn_templates_sts_session",
+            )
+            aws_credentials = assumed_role_object["Credentials"]
+            s3 = boto3.client(
+                "s3",
+                region_name=credential_region,
+                aws_access_key_id=aws_credentials.get("AccessKeyId"),
+                aws_secret_access_key=aws_credentials.get("SecretAccessKey"),
+                aws_session_token=aws_credentials.get("SessionToken"),
+            )
+
+        except ClientError:
+            print("Warning: non authorized in region '{0}', skipping".format(credential_region))
+            pass
+    else:
+        s3 = boto3.client("s3")
+    return s3
+
+
+def _get_bucket_name(args, region):
+    return region + "-aws-parallelcluster" if not args.bucket else args.bucket
+
+
+def main(args):
+
+    for _region in args.regions:
+        s3 = _create_s3_client(_region)
+        bucket1_name = _get_bucket_name(args, _region)
+
+        print("Listing cookbook for region ({0})".format(_region))
+        _aws_ls(s3, _region, bucket1_name, _COOKBOOKS_DIR + "/" + args.full_name + ".tgz")
+
+    if len(_ls_error_array) > 0 and not args.override:
+        print("We know the cookbook archives are already there, in this round we need to upload the .date files!")
+        print("Failed to push cookbook, already present for region " + " ".join(_ls_error_array))
+        exit(1)
+    elif len(_ls_error_array) > 0 and args.override:
+        print("Some or all of the cookbook archives are already there but OVERRIDE=true")
+
+    for _region in args.regions:
+        s3 = _create_s3_client(_region)
+        bucket1_name = _get_bucket_name(args, _region)
```

Comment:
I'd avoid the number in the variable name, we could call it just "bucket_name"

Reply:
sure, ok